File: frontend/src/types/property.ts
export interface PropertyImageOut {
  id: string;
  image_url: string;
  sort_order: number;
  is_cover: boolean;
}

export interface PropertyOut {
  id: string;
  property_code: string;
  name: string;
  building_name: string;
  address: string;
  latitude: number;
  longitude: number;
  district: string;
  area: string;
  nearest_bts: string | null;
  nearest_mrt: string | null;
  bedroom_count: number;
  bathroom_count: number;
  size_sqm: number;
  monthly_rent: number;
  deposit_months: number | null;
  status: string;
  available_date: string | null;
  pet_allowed: boolean;
  contact_person: string;
  contact_line: string | null;
  contact_phone: string;
  description: string;
  internal_note: string | null;
  tags: string[];
  created_by: string;
  assigned_agent_id: string | null;
  created_at: string;
  updated_at: string;
  images: PropertyImageOut[];
}

export interface PropertyListOut {
  id: string;
  property_code: string;
  name: string;
  building_name?: string;
  address: string;
  latitude?: number;
  longitude?: number;
  district: string;
  area: string;
  nearest_bts?: string | null;
  nearest_mrt?: string | null;
  bedroom_count: number;
  bathroom_count: number;
  size_sqm: number;
  monthly_rent: number;
  status: string;
  pet_allowed?: boolean;
  contact_person?: string;
  contact_line?: string | null;
  contact_phone?: string;
  tags: string[];
  distance_meters?: number;
  created_by: string;
  assigned_agent_id: string | null;
  created_at: string;
  updated_at?: string;
  images: PropertyImageOut[];
}

export interface PropertyCreate {
  property_code: string;
  name: string;
  building_name?: string;
  address: string;
  latitude: number;
  longitude: number;
  district: string;
  area: string;
  nearest_bts?: string;
  nearest_mrt?: string;
  bedroom_count: number;
  bathroom_count: number;
  size_sqm: number;
  monthly_rent: number;
  deposit_months?: number;
  status?: string;
  available_date?: string;
  pet_allowed?: boolean;
  contact_person: string;
  contact_line?: string;
  contact_phone: string;
  description?: string;
  internal_note?: string;
  tags?: string[];
  assigned_agent_id?: string;
}

export interface PropertyUpdate {
  property_code?: string;
  name?: string;
  building_name?: string;
  address?: string;
  latitude?: number;
  longitude?: number;
  district?: string;
  area?: string;
  nearest_bts?: string;
  nearest_mrt?: string;
  bedroom_count?: number;
  bathroom_count?: number;
  size_sqm?: number;
  monthly_rent?: number;
  deposit_months?: number;
  status?: string;
  available_date?: string;
  pet_allowed?: boolean;
  contact_person?: string;
  contact_line?: string;
  contact_phone?: string;
  description?: string;
  internal_note?: string;
  tags?: string[];
  assigned_agent_id?: string;
}

export interface PropertyFilter {
  search?: string;
  status?: string;
  min_price?: number;
  max_price?: number;
  min_bedrooms?: number;
  max_bedrooms?: number;
  district?: string;
  assigned_agent_id?: string;
  sort_by?: string;
  sort_order?: string;
  page?: number;
  page_size?: number;
  lat?: number;
  lng?: number;
  radius_meters?: number;
}

File: backend/app/schemas/property.py
from typing import Optional
from datetime import datetime, date
from uuid import UUID

from pydantic import BaseModel, Field


class PropertyCreate(BaseModel):
    property_code: str = Field(..., max_length=50)
    name: str = Field(..., max_length=500)
    building_name: Optional[str] = None
    address: Optional[str] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    district: Optional[str] = None
    area: Optional[str] = None
    nearest_bts: Optional[str] = None
    nearest_mrt: Optional[str] = None
    bedroom_count: Optional[int] = None
    bathroom_count: Optional[int] = None
    size_sqm: Optional[float] = None
    monthly_rent: Optional[int] = None
    deposit_months: Optional[int] = None
    status: str = "available"
    available_date: Optional[date] = None
    pet_allowed: bool = False
    contact_person: Optional[str] = None
    contact_line: Optional[str] = None
    contact_phone: Optional[str] = None
    description: Optional[str] = None
    internal_note: Optional[str] = None
    tags: Optional[list[str]] = None
    assigned_agent_id: Optional[UUID] = None


class PropertyUpdate(BaseModel):
    property_code: Optional[str] = Field(None, max_length=50)
    name: Optional[str] = Field(None, max_length=500)
    building_name: Optional[str] = None
    address: Optional[str] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    district: Optional[str] = None
    area: Optional[str] = None
    nearest_bts: Optional[str] = None
    nearest_mrt: Optional[str] = None
    bedroom_count: Optional[int] = None
    bathroom_count: Optional[int] = None
    size_sqm: Optional[float] = None
    monthly_rent: Optional[int] = None
    deposit_months: Optional[int] = None
    status: Optional[str] = None
    available_date: Optional[date] = None
    pet_allowed: Optional[bool] = None
    contact_person: Optional[str] = None
    contact_line: Optional[str] = None
    contact_phone: Optional[str] = None
    description: Optional[str] = None
    internal_note: Optional[str] = None
    tags: Optional[list[str]] = None
    assigned_agent_id: Optional[UUID] = None


class PropertyImageOut(BaseModel):
    id: UUID
    property_id: UUID
    image_url: str
    sort_order: int
    is_cover: bool
    created_at: datetime

    model_config = {"from_attributes": True}


class PropertyOut(BaseModel):
    id: UUID
    property_code: str
    name: str
    building_name: Optional[str] = None
    address: Optional[str] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    district: Optional[str] = None
    area: Optional[str] = None
    nearest_bts: Optional[str] = None
    nearest_mrt: Optional[str] = None
    bedroom_count: Optional[int] = None
    bathroom_count: Optional[int] = None
    size_sqm: Optional[float] = None
    monthly_rent: Optional[int] = None
    deposit_months: Optional[int] = None
    status: str
    available_date: Optional[date] = None
    pet_allowed: bool = False
    contact_person: Optional[str] = None
    contact_line: Optional[str] = None
    contact_phone: Optional[str] = None
    description: Optional[str] = None
    tags: Optional[list[str]] = None
    created_by: UUID
    assigned_agent_id: Optional[UUID] = None
    images: list[PropertyImageOut] = []
    created_at: datetime
    updated_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


class PropertyListOut(BaseModel):
    id: UUID
    property_code: str
    name: str
    building_name: Optional[str] = None
    address: Optional[str] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    district: Optional[str] = None
    area: Optional[str] = None
    nearest_bts: Optional[str] = None
    nearest_mrt: Optional[str] = None
    bedroom_count: Optional[int] = None
    bathroom_count: Optional[int] = None
    size_sqm: Optional[float] = None
    monthly_rent: Optional[int] = None
    status: str
    pet_allowed: bool = False
    contact_person: Optional[str] = None
    contact_line: Optional[str] = None
    contact_phone: Optional[str] = None
    tags: Optional[list[str]] = None
    distance_meters: Optional[float] = None
    created_by: UUID
    assigned_agent_id: Optional[UUID] = None
    images: list[PropertyImageOut] = []
    created_at: datetime
    updated_at: Optional[datetime] = None

    model_config = {"from_attributes": True}
